Added Command to create new worksheet

alekseiAlefirov opened the pull request:
Issue: https://github.com/scalameta/metals-vscode/issues/179
(PR for the vscode extension: https://github.com/scalameta/metals-vscode/pull/183)

From the explorer context menu:
![new-worksheet-explorer](https://user-images.githubusercontent.com/10850363/72922213-efe82a00-3d4c-11ea-98ce-a7e6f807e46d.gif)

From Command Palette:
![new-worksheet-command-palette](https://user-images.githubusercontent.com/10850363/72922228-f4acde00-3d4c-11ea-8ea2-69e179c51a42.gif)

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala:
```diff
@@ -1249,6 +1250,24 @@ class MetalsLanguageServer(
+        val args = params.getArguments.asScala
+        def f(dir: URI, name: String) = Future {
+          val path = dir.toString.toAbsolutePath.resolve(name + ".worksheet.sc")
+          //TODO: report failure, like file already exists?
```
Maybe we can just use `client.showMessage`?

alekseiAlefirov replied:
Done + `scribe.error` added

alekseiAlefirov replied:
![Screenshot from 2020-01-24 17-10-31](https://user-images.githubusercontent.com/10850363/73084220-dcfc6380-3ecc-11ea-8437-bd2b2d020c90.png)

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala:
```diff
@@ -1249,6 +1250,24 @@ class MetalsLanguageServer(
             val msg = s"Invalid arguments: $args. Expecting: $argExample"
             Future.failed(new IllegalArgumentException(msg)).asJavaObject
         }
+      case ServerCommands.NewScalaWorksheet() =>
```
Would it make sense to move it to a CommandProvider?

alekseiAlefirov replied:
While body of this method deserves (imho) being refactored to another file/class, it probably worth to be done in another PR

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala:
```diff
@@ -1249,6 +1250,24 @@ class MetalsLanguageServer(
         }
+      case ServerCommands.NewScalaWorksheet() =>
+        val args = params.getArguments.asScala
+        def f(dir: URI, name: String) = Future {
```
We should use a better name than just `f`

alekseiAlefirov replied:
`createWorksheet` now

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/ServerCommands.scala:
```diff
@@ -130,6 +130,17 @@ object ServerCommands {
+  val NewScalaWorksheet = new Command(
+    "new-scala-worksheet",
+    "Create new scala worksheet",
+    """|Create and open new scala worksheet file,
```
Nitpick, maybe without the comma and in one line?

```
    """|Create and open new scala worksheet file at the argument location with the argument + ".worksheet.sc" name.
```

alekseiAlefirov replied:
:+1:

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala:
```diff
@@ -1249,6 +1250,31 @@ class MetalsLanguageServer(
             val msg = s"Invalid arguments: $args. Expecting: $argExample"
             Future.failed(new IllegalArgumentException(msg)).asJavaObject
         }
+      case ServerCommands.NewScalaWorksheet() =>
```
Also, I think we can start with new worksheet, but even create a `WizardProvider` and `NewCommands` that would generalize over creating templates.

olafurpg commented on tests/unit/src/test/scala/tests/NewScalaWorksheetLspSuite.scala:
```diff
@@ -0,0 +1,25 @@
+import scala.meta.internal.metals.ServerCommands
+import scala.meta.internal.metals.MetalsEnrichments._
+
+object NewScalaWorksheetLspSuite extends BaseLspSuite("new-scala-worksheet") {
```
Heads up, when rebasing on top of master you will need to change this `object` into a `class`. See #1277

alekseiAlefirov commented:
Added command to create new class/object/trait file:
![NewScalaClass](https://user-images.githubusercontent.com/10850363/73384285-41536480-42cb-11ea-8360-04630ef6d4db.gif)

olafurpg commented:
That looks amazing! What do you think about having all the options in the same quickopen? Worksheet/class/object/trait/..

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala:
```diff
@@ -1262,6 +1260,55 @@ class MetalsLanguageServer(
+            val msg = s"Invalid arguments: $args."
+            Future.failed(new IllegalArgumentException(msg))
+        }).asJavaObject
+      case ServerCommands.NewScalaClass() =>
```
Is it possible to może most of the logic to the new provider you created?

tgodzik replied:
Te code grew a bit and we need to reduce the amount of code in MetalsLanguageServer I think.

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala:
```diff
@@ -1262,6 +1260,55 @@ class MetalsLanguageServer(
+      case ServerCommands.NewScalaWorksheet() =>
+        val args = params.getArguments.asScala
+        (args match {
+          case Seq(JsonString.Option(directory), JsonString(name)) =>
```
Can the parameters just contain a single object that we would translate to a class using `JsonParser`? We wouldn't need to do the crazy unapplies.

And we could use the kind parameter for everything. There wouldn't need to be separate logic for creating worksheets and classes.

alekseiAlefirov replied:
@tgodzik 
Maybe it's worth to create full `metals/newScalaFile` endpoint method? I'm not really sure which one is more appropriate in what situations.

tgodzik replied:
I think a command is better since it can also work for other clients. Although not 100 % sure

ckipp01 replied:
I'd be in favor of a `newScalaFile` endpoint since looking at it, there is a lot that is shared. They can also both share the same params and just have kind be either trait, object, class, or worksheet.

alekseiAlefirov replied:
Made one new command `new-scala-file` instead of previous two, and `MetalsNewScalaFileParams` case class with all it's needed arguments.

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala:
```diff
@@ -1917,4 +1964,19 @@ class MetalsLanguageServer(
     })
   }
 
+  private object JsonString {
```
Overall, I am not sure we need the unapplies here. You can just as easily match on JsonString.

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala:
```diff
@@ -1917,4 +1964,19 @@ class MetalsLanguageServer(
 
+  private object JsonString {
+    def unapply(json: Any): Option[String] = json match {
+      case p: JsonPrimitive if p.isString() => Some(p.getAsString())
```
Why you have: `p: JsonPrimitive if p.isString()` and below you can match on JsonString?

alekseiAlefirov replied:
when I match on JsonString below, I use this exactly method `unapply`.

alekseiAlefirov commented:
Made one new command `new-scala-file` instead of previous two.

tgodzik commented:
@alekseiAlefirov `tests.AddPackageLspSuite` is failing due to changes in the PR, it's probably something simple. And you should also run `scalafixAll` on the codebase. I will review shortly

alekseiAlefirov commented:
Oh, yes. I forgot to mention that I moved providing package declaration from `didOpen` to the new command. So it won't happen when regular new file is created. This seems to make sense, as user either uses a template provider, or go raw to set everything by themselves. IDEA does it the same way. Hope that's OK.

Yep, and some tests to provide package object is failing. So I'm thinking about adding `package object` to the command as well. What do you think?

olafurpg commented:
>  I'm thinking about adding package object to the command as well

That sounds like a great idea!

olafurpg commented:
I think it would be nice to keep the current behavior where the package gets auto-inserted when you create a new file even if it's not through the "New scala file" command.

alekseiAlefirov commented:
Okay, there's a bit of difficulty to request apply edits both from the new `new-scala-file` command and on `didOpen` (sends package declaration now): VSCode does not apply edits until the text document is open, hence `didOpen` is triggered with empty `TextDocument`,  regardless if there have been edits to apply (btw this behavior is not reflected in `TestingClient`). So for the `new-scala-file` there going to be two edits apply, conflicting by their content and (possibly) their placement in the file.
To deal with it we can:
- Do not provide package on the the new command executed.
So the order is: on `new-scala-file` edits with empty class templates requested -> client opens document -> edits with package info requested -> client inserts package info at start of the document, where empty class is already inserted. The question is, is it guaranteed that the client receives requests in the same order, that they have been sent? If not (which may result in wrong order of the new empty class declaration and the package declaration), it can be probably fixed by not requesting the empty class edit applied, but sending this edit as a response to initial request, and then manually applied (or. to be more accurate, queued to be applied) on client side before the document is open.
Also, this approach assumes though quite default, but specific handling from the client (request to execute command and then open the new document). Also, I am not sure, if it's general for all LSP clients to have this `applyEdits` and `openTextDocument` order behavior, as VSCode does.
- Provide package with the new command and save the new file by the server, not sending request to apply edits. This seems a bit out of line with LSP-logic and inconsistent with `applyEdits` request sent, which is already there on `didOpen`. Though at the same time, the new file is already created by the server side (at this moment in this PR), and manipulations like this are already being performed at some places in Metals.

Just my thoughts out loud :thinking:, there probable have been and to be more, but that's what I've gathered up at the moment. After writing this down, I'm more about just writing the whole new content  for the new class (package object, etc.) on the server side. This seems simple and expects minimum behavior from a client.

alekseiAlefirov commented:
- fixed providing package for regular empty scala file
- added creation of new package object;
- template text now inserted on server site.
![new-package-object](https://user-images.githubusercontent.com/10850363/73766724-be248980-4776-11ea-8f4b-2d36d96b2f81.gif)

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/ServerCommands.scala:
```diff
@@ -199,6 +216,13 @@ object ServerCommands {
 }
+
+case class MetalsNewScalaFileParams(
+    @Nullable directory: String,
```
We should be able to just pack it into Option? Or is it a gson limitation?

alekseiAlefirov replied:
I just followed other params case class declaration, like `MetalsStatusParams`. I would guess, it's a lsp4j limitation, not sure. Maybe @olafurpg  can comment on this ?
Then I do `Option(directory)`, yes.

olafurpg replied:
We're using gson to serialize to JSON and I think you need custom gson adapters to support `Option[T]`

olafurpg replied:
The `@Nullable` annotation isn't required, but I used it to document field where I would have preferred to use `Option[T]`

tgodzik commented:
@olafurpg what do you think about the current state?

olafurpg commented on metals/src/main/scala/scala/meta/internal/metals/NewFilesProvider.scala:
```diff
@@ -0,0 +1,96 @@
+      createPackageObject(directory)
+    case "worksheet" =>
+      createWorksheet(directory, name)
+    case _ => Future.failed(new IllegalArgumentException)
```
```suggestion
    case invalid => Future.failed(new IllegalArgumentException(invalid))
```

alekseiAlefirov commented:
> The implementation looks good to me  Great work @alekseiAlefirov 

Thanks :)
 
> What is the behavior now when users create new files through the file explorer?

![new-package-object](https://user-images.githubusercontent.com/10850363/73944327-af62e180-48f2-11ea-9c94-917c9cde40fb.gif)

olafurpg commented:
@alekseiAlefirov what is the behavior if the user runs "File: new file"?

alekseiAlefirov commented:
> @alekseiAlefirov what is the behavior if the user runs "File: new file"?

It's the same as in master: empty `.scala` files get package declaration, and `package.scala` ones get package object declaration additionally.

ckipp01 commented:
Really excited to give this a try! Great job @alekseiAlefirov! 🎉

alekseiAlefirov commented:
@ckipp01 thank you :smile:.
Still, there's client PR to get approved and merged
https://github.com/scalameta/metals-vscode/pull/183
